02_bubble/marksomething/main_test.go

```diff
@@ -0,0 +1,96 @@
+package main
+
+import (
+	"bytes"
+	"fmt"
+	"reflect"
+	"testing"
+)
+
+func Test(t *testing.T) {
+
+	algorithms := []struct {
+		desc string
+		fn   func([]int) []int
+	}{
+		{"Bubble", bubble},
+		{"Insertion", insertion},
+		{"quick", quick},
+	}
+
+	testCases := []struct {
+		desc     string
+		arg      []int
+		expected []int
+	}{
+		{
+			desc:     "Basic",
+			arg:      []int{3, 2, 1, 5, 3, 2, 1, 7, 6},
+			expected: []int{1, 1, 2, 2, 3, 3, 5, 6, 7},
+		},
+		{
+			desc:     "NegativeValues",
+			arg:      []int{3, -2, 1, 5, -3, 2, 1, -7},
+			expected: []int{-7, -3, -2, 1, 1, 2, 3, 5},
+		},
+		{
+			desc:     "SingleEntry",
+			arg:      []int{1},
+			expected: []int{1},
+		},
+		{
+			desc:     "Empty",
+			arg:      []int{},
+			expected: []int{},
+		},
+	}
+	for _, algorithm := range algorithms {
+		for _, tC := range testCases {
+			testName := fmt.Sprintf("%v-%v", algorithm.desc, tC.desc)
+			alg := algorithm
+			testCase := tC
+			t.Run(testName, func(t *testing.T) {
+				actual := alg.fn(testCase.arg)
+				expected := testCase.expected
+				if !reflect.DeepEqual(actual, expected) {
```

Comment:
at this stage I'd use testify/assert and go with `assert.Equal(t, expected, actual)` (it's unit testing convention to have expected as first arg) - in case of failure testify prints a nice diff....

Reply:
Thanks for this, that's much nicer, was having a bad time trying to find a decent way to get Printf to format for me :)